R/derive_vars_extreme_date.R

```diff
@@ -0,0 +1,154 @@
+#'
+#' @param new_var name of the extreme date derived e.g.treatment end date `TRTEDTM`
+#'
+#' @inheritParams impute_dtc
+#'
+#' @details For each group (with respect to the variables specified for the
+#'   `by_vars` parameter) the first observation (with respect to the order
+#'   specified for the `order` parameter) is included in the output dataset.
+#'
+#' @author Stefan Bundfuss, Teckla Akinyi
+#'
+#' @return The input dataset with extreme date variable added e.g. `TRTEDTM`
+#'
+#' @export
+#'
+#' @keywords adsl timing derivation
+#'
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("ex")
+#' data("dm")
+#'
+#'dm %>%
+#'  derive_vars_extreme_date(subject_keys=vars(STUDYID,USUBJID),
+#'                           dataset_date = ex,
+#'                           filter_dataset_date = EXDOSE>0,
+#'                           dtc = EXSTDTC,
+#'                           order = vars(EXSTDTC,EXSEQ),
+#'                           new_var=TRTEDTM,
+#'                           date_imputation = "first",
+#'                           time_imputation = "first",
+#'                           flag_imputation = "AUTO",
+#'                           min_dates = NULL,
+#'                           max_dates = NULL,
+#'                           mode="LAST") %>%
+#'   select(USUBJID, TRTEDTM)
+derive_vars_extreme_date <- function(dataset,
+                                     subject_keys =vars(STUDYID,USUBJID),
+                                     dataset_date,
+                                     filter_dataset_date = NULL,
+                                     dtc,
+                                     order,
+                                     new_var,
+                                     date_imputation,
+                                     time_imputation,
+                                     flag_imputation = "AUTO",
+                                     min_dates = NULL,
+                                     max_dates = NULL,
+                                     mode) {
+
+  new_var <- assert_symbol(enquo(new_var))
+  dtc <- assert_symbol(enquo(dtc))
+  assert_data_frame(dataset, subject_keys)
+  assert_data_frame(dataset_date, required_vars = quo_c(subject_keys, order))
+  assert_character_scalar(mode, values = c("first", "last"), case_sensitive = F)
+
+  filter_dataset_date <- assert_filter_cond(enquo(filter_dataset_date), optional = TRUE)
+  filter_dataset_date <- enquo(filter_dataset_date)
+
+  assert_character_scalar(flag_imputation, values = c("auto", "both", "date", "time", "none"),
+                          case_sensitive = FALSE)
+
+  #Issue warning that new_var should be DTM
+  if (str_detect(rlang::quo_name(new_var), "DTM", negate = T)) {
+    msg <- sprintf(
+      "This function derives the treatment datetime, new_var must end in DTM.")
+    warn(msg)
+  }
+
+  add <- dataset_date %>%
+    mutate(imputed_dtc = convert_dtc_to_dtm(!!dtc,
```

Comment:
Couldn't we use `derive_vars_dtm()` instead of `mutate()` and `convert_dtc_to_dtm()`?
Then we could also remove the code for deriving the imputation flags below.

Reply:
I would rather not since derive_vars_dtm uses the same mutate and convert_dtc_to_dtm meaning we would be nesting dependencies.

Reply:
I think if the code should do the same as the code of `derive_vars_dtm()`, we should not copy the code but call `derive_vars_dtm()`. Then
- it is clear from the code that the standard derivation of `derive_vars_dtm()` is used here,
- it ensures a consistent behaviour of the functions, and
- it is easier to maintain.